import os
from pathlib import Path
import pandas as pd

# load samples into table

configfile : "./config/config.yaml"
samples = pd.read_csv(config["samples"], index_col = "sample", sep ='\t')

has_reference = bool(config.get("ref_genome", "")) and bool(config.get("annotation", ""))

final_output = []
if has_reference:
    final_output.append(f"{config['output_dir_path']}/featurecounts/gene_counts.txt")
else:
    final_output.extend(
        expand(f"{config['output_dir_path']}/quant/trinity_kallisto/{{sample}}", sample=samples.index) +
        expand(f"{config['output_dir_path']}/quant/trinity_rsem/{{sample}}.isoforms.results", sample=samples.index) +
        [f"{config['output_dir_path']}/trinity_out/Trinity.fasta"]  
    )


rule all:
    input:
        f"{config['output_dir_path']}/qc/multiqc_report.html",
        *final_output

include: "rules/fastp.smk"
include: "rules/multiqc.smk"
include: "rules/star.smk"
include: "rules/featurecounts.smk"
include: "rules/denovo.smk"
include: "rules/quant_RSEM_Kallisto.smk"